### PR TITLE
Feature/manager

### PR DESCRIPTION
**Pull request by 244372610**
manager module support
## Description
1、add manager support for dynamic config
2、optimized some code details

## Related Issue
 #202

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [ ] Code compiles correctly.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by yangjun289519474 — APPROVED** [review]

File: manager/apis/v2/handler/scheduler_instance.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+// @Tags SchedulerInstance
+// @Accept  json
+// @Produce  json
+// @Param instance body types.SchedulerInstance true "Scheduler instance"
+// @Success 200 {object} types.SchedulerInstance
+// @Failure 400 {object} HTTPError
+// @Failure 404 {object} HTTPError
+// @Failure 500 {object} HTTPError
+// @Router /schedulerinstances [post]
+func (handler *Handler) AddSchedulerInstance(ctx *gin.Context) {
```
Comment (gaius-qi) [inline manager/apis/v2/handler/scheduler_instance.go:27]: Rename `AddSchedulerInstance` to `CreateScheduler`.
RESTful API: image

File: manager/apis/v2/main.go [hunk 2]
```diff
@@ -32,13 +32,40 @@ func main() {
...
+			configs := api.Group("/schedulerclusters")
+			configs.POST("", handler.AddSchedulerCluster)
+			configs.DELETE(":id", handler.DeleteSchedulerCluster)
+			configs.POST(":id", handler.UpdateSchedulerCluster)
+			configs.GET(":id", handler.GetSchedulerCluster)
+			configs.GET("", handler.ListSchedulerClusters)
+		}
+
+		{
+			configs := api.Group("/schedulerinstances")
```
Comment (gaius-qi) [inline manager/apis/v2/main.go:45]: Use `/scheduler` path.

File: manager/apis/v2/main.go [hunk 3]
```diff
@@ -32,13 +32,40 @@ func main() {
...
+			configs.POST("", handler.AddSchedulerInstance)
```
Comment (gaius-qi) [inline manager/apis/v2/main.go:46]: Remove `Instance`.

File: manager/apis/v2/main.go [hunk 4]
```diff
@@ -32,13 +32,40 @@ func main() {
...
+			configs.DELETE(":id", handler.DeleteSchedulerInstance)
+			configs.POST(":id", handler.UpdateSchedulerInstance)
+			configs.GET(":id", handler.GetSchedulerInstance)
+			configs.GET("", handler.ListSchedulerInstances)
+		}
+
+		{
+			configs := api.Group("/cdnclusters")
```
Comment (gaius-qi) [inline manager/apis/v2/main.go:54]: Use `/cdn/clusters` path.

File: manager/apis/v2/main.go [hunk 5]
```diff
@@ -32,13 +32,40 @@ func main() {
...
+			configs.POST("", handler.AddCdnCluster)
+			configs.DELETE(":id", handler.DeleteCdnCluster)
+			configs.POST(":id", handler.UpdateCdnCluster)
+			configs.GET(":id", handler.GetCdnCluster)
+			configs.GET("", handler.ListCdnClusters)
+		}
+
+		{
+			configs := api.Group("/cdninstances")
```
Comment (gaius-qi) [inline manager/apis/v2/main.go:63]: Use `/cdn/instances` path and rename `Cdn` to `CDN`.

File: manager/apis/v2/main.go [hunk 6]
```diff
@@ -32,13 +32,40 @@ func main() {
...
+			configs.POST("", handler.AddCdnInstance)
+			configs.DELETE(":id", handler.DeleteCdnInstance)
+			configs.POST(":id", handler.UpdateCdnInstance)
```
Comment (gaius-qi) [inline manager/apis/v2/main.go:66]: Use `PATCH` method.

File: manager/apis/v2/types/schema.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
+package types
+
+type SchedulerCluster struct {
+	ClusterId       string `json:"cluster_id,size:63" binding:"omitempty"`
```
Comment (gaius-qi) [inline manager/apis/v2/types/schema.go:4]: Why use size 63?

Varchar  is 255 is useful value.

https://stackoverflow.com/questions/2340639/why-historically-do-people-use-255-not-256-for-database-field-magnitudes

File: manager/apis/v2/types/schema.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+	SchedulerConfig string `json:"scheduler_config,size:4095" binding:"required"`
+	ClientConfig    string `json:"client_config,size:4095" binding:"required"`
```
Comment (gaius-qi) [inline manager/apis/v2/types/schema.go:6]: JSON use lower camel case.

File: manager/apis/v2/types/schema.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+	Version         int64  `json:"version" binding:"omitempty"`
+	Creator         string `json:"creator,size:31" binding:"omitempty"`
+	Modifier        string `json:"modifier,size:31" binding:"omitempty"`
+	CreatedAt       string `json:"created_at" binding:"omitempty"`
+	UpdatedAt       string `json:"updated_at" binding:"omitempty"`
+}
+
+type SchedulerClusterUri struct {
+	ClusterId string `uri:"id" binding:"required"`
```
Comment (gaius-qi) [inline manager/apis/v2/types/schema.go:15]: Rename `Id` to `ID`.

File: manager/config/config.go [hunk 10]
```diff
@@ -28,21 +29,27 @@ type MysqlConfig struct {
...
-	Memory *MemoryConfig `yaml:"memory", omitempty`
+type HostService struct {
+	Skyline *SkylineService `yaml:"skyline"`
+}
+
+type SkylineService struct {
+	Domain    string `yaml:"domain"`
+	AppName   string `yaml:"app-name"`
+	Account   string `yaml:"account"`
+	AccessKey string `yaml:"access-key"`
```
Comment (gaius-qi) [inline manager/config/config.go:52]: JSON use lower camel case.

File: manager/config/config.go [hunk 11]
```diff
@@ -28,21 +29,27 @@ type MysqlConfig struct {
...
-	Mysql  *MysqlConfig  `yaml:"mysql",omitempty`
-	Oss    *OssConfig    `yaml:"oss",omitempty`
-	Memory *MemoryConfig `yaml:"memory", omitempty`
+type HostService struct {
+	Skyline *SkylineService `yaml:"skyline"`
+}
+
+type SkylineService struct {
+	Domain    string `yaml:"domain"`
+	AppName   string `yaml:"app-name"`
```
Comment (gaius-qi) [inline manager/config/config.go:50]: JSON use lower camel case.

File: manager/config/config.go [hunk 12]
```diff
@@ -63,10 +70,16 @@ func New() *Config {
...
-					Table:    "config_table",
 				},
-				Oss:    nil,
-				Memory: nil,
+				Oss: nil,
+			},
+		},
+		HostService: &HostService{
+			Skyline: &SkylineService{
+				Domain:    "http://xxx",
```
Comment (gaius-qi) [inline manager/config/config.go:79]: Use `example` or `foo`.

File: manager/host/host_manager.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+	Location       string
+	Idc            string
+	NetTopology    string
+}
+
+func NewDefaultHostInfo(ip, hostName string) *HostInfo {
+	return &HostInfo{
+		HostName:       hostName,
+		Ip:             ip,
+		SecurityDomain: "",
```
Comment (gaius-qi) [inline manager/host/host_manager.go:22]: The default value does not require display assignment.

File: manager/host/host_manager.go [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+		Idc:            "",
+		NetTopology:    "",
+	}
+}
+
+func (host *HostInfo) IsDefault() bool {
+	return len(host.SecurityDomain) == 0 && len(host.Idc) == 0
+}
+
+type Op struct {
```
Comment (gaius-qi) [inline manager/host/host_manager.go:33]: What is OP?

File: manager/host/skyline.go [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+func (req *QueryRequest) SetFrom(from string) {
+	req.From = from
+}
+
+func (req *QueryRequest) SetCondition(cond string) {
+	req.Condition = cond
+}
+
+type Item struct {
+	Ip                 string   `json:"ip"`
```
Comment (gaius-qi) [inline manager/host/skyline.go:55]: English abbreviations need to be uppercase.

File: manager/host/skyline.go [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+	}
+
+	url := fmt.Sprintf("%s/item/query", client.domain)
+	request, err := http.NewRequest("POST", url, bytes.NewReader(body))
+	defer request.Body.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	request.Header.Set("Content-Type", "application/json;charset=UTF-8")
```
Comment (gaius-qi) [inline manager/host/skyline.go:139]: Use a better way, without the need for explicit assignment.

File: manager/host/skyline.go [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+	request.Header.Set("account", client.account)
```
Comment (gaius-qi) [inline manager/host/skyline.go:140]: What is `account` header?
If you use a custom header, prefix it with `x-`.

File: manager/host/skyline.go [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+		client: &skylineClient{
+			domain:    config.Domain,
+			appName:   config.AppName,
+			account:   config.Account,
+			accessKey: config.AccessKey,
+		}}, nil
+}
+
+var SelectLists = []string{
+	"ip", "sn", "tags", "host_name", "security_domain", "vpc_id",
```
Comment (gaius-qi) [inline manager/host/skyline.go:177]: Use lower camel case.

File: manager/server/router.go [hunk 19]
```diff
@@ -15,13 +15,40 @@ func InitRouter(server *service.ManagerServer) (*gin.Engine, error) {
```
Comment (gaius-qi) [inline manager/server/router.go:15]: There are two identical router files.

File: pkg/rpc/cdnsystem/client/cdn_hosts_mgr.go [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"context"
+
+	"d7y.io/dragonfly/v2/pkg/rpc/manager"
+	mgClient "d7y.io/dragonfly/v2/pkg/rpc/manager/client"
+	"d7y.io/dragonfly/v2/pkg/util/net/iputils"
+)
+
+type DynamicHostsMgr struct {
```
Comment (gaius-qi) [inline pkg/rpc/cdnsystem/client/cdn_hosts_mgr.go:28]: Try to be consistent in naming, use `dyn`.

File: pkg/rpc/cdnsystem/client/cdn_hosts_mgr.go [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+func (md *DynamicHostsMgr) Get() (interface{}, error) {
+	scConfig, err := md.managerCli.GetSchedulerClusterConfig(context.Background(), &manager.GetClusterConfigRequest{
+		HostName: iputils.HostName,
+		Type:     manager.ResourceType_Scheduler,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return scConfig.CdnHosts, nil
```
Comment (gaius-qi) [inline pkg/rpc/cdnsystem/client/cdn_hosts_mgr.go:45]: Rename `CdnHosts` to `CDNHosts`.

**Comment by gaius-qi** [comment]
Need to fix lint and unit test issues before they can be merged.

File: manager/host/skyline.go (hunk 18 above)
Comment (yangjun289519474) [reply manager/host/skyline.go:177]: bind with skyline sql backend, these are key words of query

File: manager/host/skyline.go (hunk 17 above)
Comment (yangjun289519474) [reply manager/host/skyline.go:140]: specd by skyline(cmdb) header, not manager custom header

File: manager/config/config.go (hunk 10 above)
Comment (yangjun289519474) [reply manager/config/config.go:52]: bind with skyline sql backend

**Comment by gaius-qi** [comment]
LGTM

File: cdnsystem/config/config.go [hunk 22]
```diff
@@ -166,9 +166,12 @@ type BaseProperties struct {
 	// default: 3min
 	TaskExpireTime time.Duration `yaml:"taskExpireTime"`
 
-	// disk/hybrid/memory
+	// StoragePattern disk/hybrid/memory
```
Comment (gaius-qi) [inline cdnsystem/config/config.go:169]: foo

Comment (gaius-qi) [inline cdnsystem/config/config.go:169]: foo

File: pkg/rpc/cdnsystem/client/cdn_hosts_mgr.go (hunk 21 above)
Comment (244372610) [reply pkg/rpc/cdnsystem/client/cdn_hosts_mgr.go:45]: rpc生成

**Review by gaius-qi — APPROVED** [review]

**Review by yangjun289519474 — APPROVED** [review]

**Pull request merged**